### ISSUE TITLE
`simulatedAccessibilityTraversal` doesn't account for explicit/force-merging (via `MergeSemantics`)

### ISSUE THREAD

**Issue by marcianx** [flutter/flutter#135144]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Use `MergeSemantics` to wrap one or more multiple widgets that emit `Semantics(container: true)` to consolidate the semantics tree.
2. Test the semantics of the consolidated widget using `tester.semantics.simulatedAccessibilityTraversal()`.

See code further below.

### Expected results

Per @goderbauer, `simulatedAccessibilityTraversal()` should account for the force-merging to enable unit testing of `MergeSemantics`-based code. Effectively, it should be testing a `SemanticsData` tree to capture the platform accessibility tree that will be generated from the semantics tree.

### Actual results

`simulatedAccessibilityTraversal()` enumerates the semantics tree as-is, including all `SemanticsNode`s that are merged upward. It does not use the `SemanticsData` due to force-merging.

Maybe this is inherent to the API? Should it be deprecated in favor of a new API that emits `SemanticsData` (and also exposes the number of children so that we can assert when something should be a leaf node)? Is it possible to test the structure of the platform accessibility that will be generated?

### Code sample

https://dartpad.dev/?id=06937c78a65c3c3daf75eb81648b1783

<details><summary>Code sample</summary>

```dart
await tester.pumpWidget(
  MaterialApp(
    home: MergeSemantics(
      child: Semantics(
        container: true,
        label: 'Container',
        child: Semantics(
          button: true,
          container: true,
          label: 'Foo',
          child: const SizedBox(width: 10, height: 10),
        ),
      ),
    ),
  ),
);

// Fails since `simulatedAccessibilityTraversal` doesn't account
// for force-merging.
expect(
  tester.semantics.simulatedAccessibilityTraversal(),
  [containsSemantics(label: 'Container\nFoo', isButton: true)],
);
```

</details>

**Comment by marcianx** [comment]
So `containsSemantics` does test against `SemanticsData`. So effectively, this means that we want to rewrite the check as:

https://dartpad.dev/?id=d1dba6877fc8978bd319259f8e39db65

```dart
expect(
  tester.semantics
      .simulatedAccessibilityTraversal()
      .where((node) => !node.isMergedIntoParent),
  [containsSemantics(label: 'Container\nFoo', isButton: true)],
);
```

I can't help but wonder whether this filtering should be on by default. WDYT @goderbauer?

**Comment by goderbauer** [comment]
From experimenting with the provided dartpad it seems like `simulatedAccessibilityTraversal` currently traverses over a node that contains all the force-merged data as well as the individual nodes that got merged into it. This seems like a bug and `simulatedAccessibilityTraversal` should just internally filter out the nodes that got merged. This would ensure it implements the documented (and desired) behavior:

> Simulates a traversal of the currently visible semantics tree **as if by assistive technologies**.

Since assistive technologies don't get to see the merged nodes, `simulatedAccessibilityTraversal` shouldn't give it to you either.

@pdblasi-google @chunhtai WDYT?

**Comment by goderbauer** [comment]
> One interesting consequence is that we just can't rely on containsSemantics(children: []) to assert that it's a leaf node in the platform accessibility tree.

IMO, `containsSemantics` shouldn't see any children if those are actually force-merged into the parent node. This is another thing we should fix. If `matchesSemantics` has the same problem, we need to do it there as well.

**Comment by pdblasi-google** [comment]
Do assistive technologies _not_ see merged nodes? I thought that was the purpose of merging nodes, to make assistive technologies treat everything as a singular node.

**Comment by pdblasi-google** [comment]
On a re-read, I think I misunderstood your explanation @goderbauer. To phrase it a different way, we would be updating `simulatedAcessibilityTraversal` such that nothing with `isMergedIntoParent = true`, correct? 

If that's the case, then yes, that was the intended behavior so this was an oversight in the original implementation.

**Comment by chunhtai** [comment]
Yes I agree it should ignore anything nodes under the semantics node with `isMergedIntoParent = true`. We are verifying the merged result anyway when we call `matchesSemantics` or `containsSemantics`

**Comment by pdblasi-google** [comment]
@chunhtai

Just for my curiosity, do you know where they're filtered out when they're passed to the engine? I don't remember them being explicitly excluded in the engine code (which is where I got the other flag/action checks for `simulatedAccessibilityTraversal`), so I guess I assumed they'd be filtered out in the construction of the tree.

It's been a while, so it's quite possible I just missed it, just curious.

**Comment by goderbauer** [comment]
> To phrase it a different way, we would be updating simulatedAcessibilityTraversal such that nothing with isMergedIntoParent = true is included in the results, correct?

Correct. Sorry if my wording was unclear.

> Just for my curiosity, do you know where they're filtered out when they're passed to the engine?

This happens just before we send the semantics data to the engine:

https://github.com/flutter/flutter/blob/4d8e390c1be6237ab45df9c57b87e50e499056db/packages/flutter/lib/src/semantics/semantics.dart#L2684-L2690

(getSemanticsData returns the data after the merge and we then omit sending the child nodes)

**Comment by pdblasi-google** [comment]
This should be pretty easy to resolve by updating `_isImportantForAccessibliity` to check for `isMergedIntoParent`, which would let the rest of the traversal/logic go untouched.
